File: sample_geometry.R
library(tidyverse)
source("utility_functions.R")
#ch3

#ex3.1
X<-matrix(c(9,5,1,1,3,2),ncol=2,nrow = 3,byrow = F)

X %>% as_tibble() %>% ggplot(aes(x=V1,y=V2)) + 
    geom_point() +
    geom_point(aes(x=mean(V1),y=mean(V2)),color="red")  #mean coordinate

# p=2 vector in n=3 space
library(plotly)
plot_ly(data = as.data.frame(t(X)), x=~V1,y=~V2, z=~V3, type="scatter3d") 
?plot_ly






# 3.9

X<-matrix(c(12,18,14,20,16,17,20,16,18,19,29,38,30,38,35),ncol=3,nrow = 5,byrow = F)
X
xbar <- colMeans(X)
xbar
xbar<-ones(nrow(X))%*%xbar #as n by p matrix
Z <- X-xbar #mean corrected
Z
#notice that col 3 is col1+col2
# therefore col1+col2 - col3 = 0
# ie columns are linearly dependent
a<-matrix(c(1,1,-1), nrow = 3,ncol=1)
a
Z%*%a # = 0 

S <- my_covmatrix(Z)
det(S) # the generalized sample variance is basically zero 
S%*%a # = 0 , ie a can be rescaled to be an eigenvector corresponding to eigenvalue zero


# 3.14
X<-matrix(c(9,5,1,1,3,2),ncol=2,nrow = 3,byrow = F)
c <- c(-1,2) %>% as.matrix()
b <- c(2,3) %>% as.matrix()
b
X
S<-my_covmatrix(X)
X
t(b)%*%X[1,]
t(b)%*%X[2,]
t(b)%*%X[3,]

xbar<- colMeans(X) %>% as.matrix()
t(b)%*%xbar
t(c)%*%xbar
t(b)%*%S%*%b
t(c)%*%S%*%c
t(b)%*%S%*%c


#  3.15
X<-matrix(c(1,6,8,4,2,3,3,6,3),ncol=3,nrow = 3,byrow = F)
n<-nrow(X)
xbar <- colMeans(X)
XBAR<-ones(nrow(X))%*%xbar #as n by p matrix
I<- diag(n)
J <- ones(n)%*%t(ones(n))
S <- (1/(n-1))*t(X)%*%(I-(1/n)*J)%*%X # or:
S <- (1/(n-1))*t(X)%*%(X-XBAR)
my_covmatrix(X)

#linear combinartions
b <- c(1,1,1) %>% as.matrix()
c <- c(1,2,-3) %>% as.matrix()

t(b)%*%xbar #sample mean of b'X
t(c)%*%xbar  #sample mean of c'X
t(b)%*%S%*%b #sample variance of b'X
t(c)%*%S%*%c #sample variance of c'X
t(b)%*%S%*%c  #sample covariance of b'X an c'X




cc <- matrix( c(6, -10/4, -10/4, 1.5),nrow=2,ncol=2)
cc <- matrix( c(48.68,-26,-26,14),nrow=2,ncol=2)
cc<-cc/2
solve(cc)
det(cc)
4/27


Y<-matrix(c(-3,4,5,15,16,11),nrow=3,ncol=2, byrow = F)
Y
cov(Y)
xbar<-colMeans(Y)
xbar<-ones(3)%*%t(xbar)
S<-(1/2)*t(Y)%*%(Y-xbar)
solve(cov(Y))
i<-solve(S)
my_covmatrix(Y)
mu <- c(4,14) %>% as.matrix()
ones(n)%*%t(mu)
xbar <- as.matrix(xbar)
xbar
n<-3
n*(t(xbar-mu)%*%i%*%(xbar-mu))
7/9

File: sample_geometry.py
import numpy as np


X = np.array([[9,1],
             [5,3],
             [1,2]])
X
col1 = X[:,0] 
col2 = X[:,1] 


# https://matplotlib.org/tutorials/introductory/pyplot.html

# n points in p-dimensional space
# - each column of X is a dimension
# - variability occurs in more than one direction
# - sample mean is center of gravity
# - determinant of sample covmatrix provides measure of variability


import matplotlib.pyplot as plt
plt.scatter(col1,col2,label="observations")
plt.scatter(np.mean(col1), np.mean(col2), color="r", label="sample mean")
plt.xlabel('column 1')
plt.ylabel('column 2')
plt.suptitle("n points in p-dimensional space")
plt.title("Each row is an observation, each column a variable. n=3, p=2", size="small")
plt.legend()



# p VECTORS in n-dimensional space

fig = plt.figure()
ax = fig.add_subplot(121, projection='3d')
#ax.autoscale(enable=False)
ax.set_xlim(0,9)
ax.set_ylim(0,6)
ax.set_zlim(0,6)
origin = [0], [0], [0]
# In Python3.5+, the asterisk unpacks a list
ax.quiver(*origin, X[0,:],X[1,:], X[2,:], arrow_length_ratio=0.1 )
plt.title("Quiver plot")

# as scatterplot: 
ax = fig.add_subplot(122, projection='3d')
ax.set_xlim(0,9)
ax.set_ylim(0,6)
ax.set_zlim(0,6)
ax.scatter(X[0,:],X[1,:], X[2,:])
plt.suptitle("p vectors in n-dimensional space. p=2, n=3")
plt.title("Scatter plot")


#####################

fig = plt.figure()
ax = fig.add_subplot(121, projection='3d')
#ax.autoscale(enable=False)
ax.set_xlim(0,9)
ax.set_ylim(0,6)
ax.set_zlim(0,6)
origin = [0], [0], [0]
# In Python3.5+, the asterisk unpacks a list
ax.quiver(*origin, X[0,:],X[1,:], X[2,:], arrow_length_ratio=0.1 )
plt.title("Quiver plot")


xbar = np.mean(X, axis=0) 
nrow = X.shape[0] # in R: nrow(X2)
Xbar = np.outer(np.ones(nrow),xbar)
Z1 = X - Xbar # deviation matrix

fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
#ax.autoscale(enable=False)
ax.set_xlim(0,9)
ax.set_ylim(0,6)
ax.set_zlim(0,6)
origin = [0], [0], [0]
# In Python3.5+, the asterisk unpacks a list
ax.quiver(*origin, X[0,:],X[1,:], X[2,:], arrow_length_ratio=0.1, label="x_i", color="b" )
ax.quiver(*origin, Z1[0,:],Z1[1,:], Z1[2,:], arrow_length_ratio=0.1 , label="deviation", color = "r")
ax.quiver(*origin, Xbar[0,:],Xbar[1,:], Xbar[2,:], arrow_length_ratio=0.1 , label="xbar", color = "g")
plt.title("Quiver plot")






#####################


X2 = np.array([[12,17,29],
               [18,20,38],
               [14,16,30],
               [20,18,38],
               [16,19,35]])

# X2 is a multivariate matrix with 5 observations with 3 dimensions(columns)

#axis = 0 means along the column and axis = 1 means working along the row.
xbar = np.mean(X2, axis=0) # same as colMeans(X2) in R 
type(xbar)

# visualize in 3d:

fig = plt.figure()
# Alternative form for add_subplot(111) is add_subplot(1, 1, 1)
# These are subplot grid parameters encoded as a single integer. For example, "111" means "1x1 grid
#(numrows, numcols, position)
#ax = fig.add_subplot(111, projection='3d')
ax = fig.add_subplot(121, projection='3d')
ax.scatter(xs= X2[:,0], ys = X2[:,1], zs = X2[:,2])
ax.scatter(xs=xbar[0], ys=xbar[1], zs = xbar[2] , color="r", label = "mean vector")
ax.legend()


nrow = X2.shape[0] # in R: nrow(X2)
# convert column means as n by p matrix:
Xbar = np.outer(np.ones(nrow),xbar) # dot product with 1d arrays in np
#see  https://stackoverflow.com/questions/23566515/multiplication-of-1d-arrays-in-numpy 
Xbar

# the mean corrected matrix / the deviation matrix:
Z = X2 - Xbar
Z

ax = fig.add_subplot(122, projection='3d')
ax.scatter(xs= Z[:,0], ys = Z[:,1], zs = Z[:,2])

# notice that col1+col2=col3
# therefore col1+col2 - col3 = 0
# ie columns are linearly dependent
a = np.array([1,1,-1])
Z@a # array([0., 0., 0., 0., 0.])



def my_sample_covmatrix(X):
    """
    Parameters
    ----------
    X : n by p matrix

    Returns
    -------
    Covariance matrix of sample (p by p)
    Unbiased (n-1).
    """
    
    n = X.shape[0]
    I = np.identity(n) 
    J = np.outer(np.ones(n),np.ones(n))
    S = (1/(n-1))*X.T@(I-(1/n)*J)@X
    return(S)

S = my_sample_covmatrix(X2)
S

np.linalg.det(S) # the generalized sample variance is basically zero 

S@a # = 0 , ie a can be rescaled to be an eigenvector corresponding to eigenvalue zero

np.linalg.eigvals(S)



